docs/src/components/ComponentDoc/ComponentExample/ComponentExample.js

```diff
@@ -64,36 +62,44 @@ class ComponentExample extends PureComponent {
     renderHtml: true,
   }
 
-  componentWillMount() {
-    const { examplePath } = this.props
-    this.anchorName = examplePathToHash(examplePath)
-
-    this.setState({
-      showCode: this.isActiveHash(),
-      sourceCode: this.getOriginalSourceCode(),
-    })
-  }
-
-  componentWillReceiveProps(nextProps) {
-    const { examplePath, exampleSources, location } = nextProps
-    const nextSourceCode = exampleSources[examplePath]
+  static getDerivedStateFromProps(nextProps, prevState = {}) {
+    const {
+      exampleKeys,
+      examplePath,
+      exampleSources,
+      location: { hash },
+    } = nextProps
+
+    const anchorName = examplePathToHash(examplePath)
+    const isActiveHash = anchorName === getFormattedHash(exampleKeys, hash)
+    const originalSourceCode = exampleSources[examplePath]
+
+    const state = {
+      anchorName,
+      hash,
+      isActiveHash,
+      originalSourceCode,
+    }
 
     // deactivate examples when switching from one to the next
-    if (this.isActiveHash() && this.isActiveState() && this.props.location.hash !== location.hash) {
-      this.clearActiveState()
+    if (prevState.isActiveHash && !isActiveHash) {
+      state.showCode = false
+      state.showHTML = false
     }
 
-    // for local environment
-    if (process.env.NODE_ENV !== 'production' && this.getOriginalSourceCode() !== nextSourceCode) {
-      this.setState({ sourceCode: nextSourceCode })
-    }
+    return state
   }
 
-  clearActiveState = () => {
-    this.setState({
-      showCode: false,
-      showHTML: false,
-    })
+  constructor(props) {
+    super(props)
+
+    const state = ComponentExample.getDerivedStateFromProps(props)
```

Comment:
`getDerivedStateFromProps()` will be dispatcher right after `constructor()`, we don't need this

Reply:
We use computation for state here to be able to enable source code displaying and set initial value for source code (see an assignment below).

Reply:
I understand what you mean, but I prefer separation of concerns there: if we will change something in `getDerivedStateFromProps()` it will break constructor.

Other confusion is placed around lifecycle: https://code.likeagirl.io/understanding-react-component-life-cycle-49bf4b8674de (we call `constructor()`, `getDerivedStateFromProps()` and then again `getDerivedStateFromProps()`)

```js
const originalSourceCode = props.exampleSources[props.examplePath]

this.state = {
  anchorName: examplePathToHash(props.examplePath),
  originalSourceCode,
  showCode: anchorName === props.location.hash,
  sourceCode: originalSourceCode,
}
```

Reply:
I can just remove line #93 (`...state,`).